web/shared/src/typing_status.ts

```diff
@@ -93,23 +90,25 @@ export function maybe_ping_server(
  *
  * @param {*} worker Callbacks for reaching the real world. See typing.js
  *   for implementations.
- * @param {*} new_recipient_ids The users the direct message being composed is
- *   addressed to, as a sorted array of user IDs; or `null` if no direct message
- *   is being composed anymore.
+ * @param {*} new_recipient Depends on type of message being composed. If
+ *   * Private message: The users the PM being composed is addressed to,
+ *     as a sorted array of user IDs
+ *   * Stream message: An Object containing the stream_id and topic
+ *   * No meesage is being composed: `null`
  */
 export function update(
     worker: TypingStatusWorker,
-    new_recipient_ids: number[] | null,
+    new_recipient: Recipient | null,
     typing_started_wait_period: number,
     typing_stopped_wait_period: number,
 ): void {
     if (state !== null) {
         // We need to use _.isEqual for comparisons; === doesn't work
-        // on arrays.
-        if (_.isEqual(new_recipient_ids, state.current_recipient_ids)) {
+        // on arrays and objects.
+        if (_.isEqual(new_recipient, state.current_recipient)) {
```

Comment:
I'm not sure this check is correct given topics being case-insensitive; we may need to use something like `util.same_recipient` or `util.same_stream_and_topic`.

Reply:
@timabbott 

Actually, `util.ts` has 
```
export type Recipient =
    | {type: "private"; to_user_ids?: string; reply_to: string}
    | ({type: "stream"} & StreamTopic);
```
```
function util_same_recipient(a?: Recipient, b?: Recipient): boolean { }
```
We can use them. They are a fit for our usecase.

One doubt I have is "what is our philosophy behind keeping files in `web/shared/src` and `web/src`? 
`typing_status.ts` and `util.ts` lie in those two different places.

Reply:
So `web/shared/src` is for sharing code with the React Native mobile apps -- it's intended to be self-contained so that it can be used from that other codebase.

Given that we're rewriting the mobile apps in Flutter, it's not a very high priority to maintain that distinction long term. Let me start a discussion on what the plan should be for that.

Reply:
> Let me start a discussion on what the plan should be for that.

@timabbott Did you start a discussion?

Reply:
Good question @prakhar1144. It looks like @timabbott did start a chat discussion on this last week shortly after that comment, but just forgot to add a link to it from here. The discussion is at:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/web.2Fshared.2Fsrc.20in.20zulip.2Fzulip/near/1651184

Reply:
Posted in that thread to move this forward. And yeah, sorry for forgetting to backlink the discussion after starting it.

Reply:
Okay, while working on this, I realized that the functions in `utils.ts` can't be used as it is. So just created a couple of similar functions for our check of `same_recipients`.